clients/drcachesim/tracer/tracer.cpp

```diff
@@ -211,48 +217,187 @@ static uint     tls_offs;
+    }
+    drvector_unlock(&circular_buf_queue);
+    return get;
+}
+
+static void
+trace_traverse(byte *buf_base, ssize_t size,
+               bool vaddr2paddr, bool do_write, thread_id_t tid)
+{
+    byte *mem_ref, *buf_ptr;
+    byte *pipe_start, *pipe_end;
+    pipe_start = buf_base;
+    pipe_end = pipe_start;
+    buf_ptr = buf_base + size;
+    for (mem_ref  = buf_base + buf_hdr_slots_size;
+         mem_ref  < buf_ptr;
+         mem_ref += instru->sizeof_entry()) {
+        if (vaddr2paddr) {
+            trace_type_t type = instru->get_entry_type(mem_ref);
+            if (type != TRACE_TYPE_THREAD &&
+                type != TRACE_TYPE_THREAD_EXIT &&
+                type != TRACE_TYPE_PID) {
+                addr_t virt = instru->get_entry_addr(mem_ref);
+                addr_t phys = physaddr.virtual2physical(virt);
+                DR_ASSERT(type != TRACE_TYPE_INSTR_BUNDLE);
+                if (phys != 0)
+                    instru->set_entry_addr(mem_ref, phys);
+                else {
+                    // XXX i#1735: use virtual address and continue?
+                    // There are cases the xl8 fail, e.g.,:
+                    // - vsyscall/kernel page,
+                    // - wild access (NULL or very large bogus address) by app
+                    NOTIFY(1, "virtual2physical translation failure for "
+                           "<%2d, %2d, " PFX">\n",
+                           type, instru->get_entry_size(mem_ref), virt);
+                }
+            }
+        }
+        if (do_write) {
+            // Split up the buffer into multiple writes to ensure
+            // atomic pipe writes.
+            // We can only split before TRACE_TYPE_INSTR, assuming only a few
+            // data entries in between instr entries.
+            if (instru->get_entry_type(mem_ref) == TRACE_TYPE_INSTR) {
+                if ((mem_ref - pipe_start) > ipc_pipe.get_atomic_write_size())
+                    pipe_start = atomic_pipe_write(tid, pipe_start, pipe_end);
+                // Advance pipe_end pointer
+                pipe_end = mem_ref;
+            }
+        }
+    }
+    if (do_write) {
+        // Write the rest to pipe
+        // The last few entries (e.g., instr + refs) may exceed the atomic write size,
+        // so we may need two writes.
+        if ((buf_ptr - pipe_start) > ipc_pipe.get_atomic_write_size())
+            pipe_start = atomic_pipe_write(tid, pipe_start, pipe_end);
+        if ((buf_ptr - pipe_start) > (ssize_t)buf_hdr_slots_size)
+            atomic_pipe_write(tid, pipe_start, buf_ptr);
+    }
+}
+
+static void
+sideline_run(void *arg)
+{
+    sideline_arg_t *sa = (sideline_arg_t *)arg;
+    buf_entry_t entry;
+    bool profiling = true;
+    bool offline = op_offline.get_value();
+    /* consume the queue and write data to file */
+    do {
+        /* quick racy check without locking */
+        while (queue_get == queue_put) {
+            /* the queue is empty, waiting */
```

Comment:
Similarly to the queue being full, it would be better to have a proper kernel-involved wait here.